packages/ui/src/composables/useValidation.ts

```diff
@@ -66,12 +69,28 @@ export const useValidation = (
     const rules = flatten(props.rules)
 
     normalizeValidationRules(rules, props.modelValue)
-      .forEach((validationResult: boolean | string) => {
+      .forEach((validationResult: validationResult) => {
         if (isString(validationResult)) {
           errorMessages = [...errorMessages, validationResult]
           error = true
         } else if (validationResult === false) {
           error = true
+        } else if (validationResult instanceof Promise) {
+          innerLoadingState.value = true
+
+          validationResult
+            .then(promisedValidationResult => {
```

Comment:
You forget to handle boolean result here. Let's have some recursive handle function here.